lib/security/JweToken.ts

```diff
@@ -14,6 +15,49 @@ type EncryptDelegate = (data: Buffer, jwk: PublicKey) => Promise<Buffer>;
  * This class hides the JOSE and crypto library dependencies to allow support for additional crypto algorithms.
  */
 export default class JweToken extends JoseToken {
+
+  // used for verification if a JSON Serialized JWS was given
+  private readonly protected: string | undefined;
+  private unprotected: {[key: string]: any} | undefined;
+  private readonly encrypted_key: string | undefined;
+  private readonly iv: string | undefined;
+  private readonly tag: string | undefined;
+  private readonly aad: string | undefined;
+  private readonly isFlattenedJSONSerialized: boolean;
+
+  public constructor (content: string | object, protected cryptoFactory: CryptoFactory) {
+    super(content, cryptoFactory);
+    const jsonContent: any = content;
+    this.isFlattenedJSONSerialized = false;
+    if (typeof jsonContent === 'object' &&
+    'ciphertext' in jsonContent && typeof jsonContent.ciphertext === 'string' &&
+    'iv' in jsonContent && typeof jsonContent.iv === 'string' &&
+    'tag' in jsonContent && typeof jsonContent.tag === 'string' &&
+    (('protected' in jsonContent && typeof jsonContent.protected === 'string') ||
+    ('unprotected' in jsonContent && typeof jsonContent.unprotected === 'object'))) {
+      if ('protected' in jsonContent) {
+        this.protected = jsonContent.protected;
+      }
+      if ('unprotected' in jsonContent) {
+        this.unprotected = jsonContent.unprotected;
+      }
+      this.iv = jsonContent.iv;
+      this.tag = jsonContent.tag;
+      this.aad = jsonContent.aad;
+      if ('recipients' in jsonContent) {
+        // TODO: General JWE JSON Serialization
+      } else if ('encrypted_key' in jsonContent && typeof jsonContent.encrypted_key === 'string') {
+        // Flattened JWE JSON Serialization
+        this.encrypted_key = jsonContent.encrypted_key;
+        if ('header' in jsonContent && typeof jsonContent.header === 'object') {
+          this.unprotected = Object.assign(this.unprotected, jsonContent.header);
+        }
+        this.isFlattenedJSONSerialized = true;
```

Comment:
Sanity: is it the expected behavior that a `JweToken` is still constructed when neither `recipients` nor `encrypted_key` are present in jsonContent?

Reply:
There are a couple convoluted fallback behaviors here. A JSON serialized JWE does require either `recipients` or `encrypted_key`, and it should fail to parse a JWE if either of those are missing, however this could also just be a JSON object that happens to look like a JWE, intended to be wrapped in an actual JWE, in which the default JoseToken constructor behavior of stringifying it as `content` is correct.

One of the problems with the JWE/JWS code is that there are two distinct behaviors between "I've been given random content, and will probably be expect to sign/encrypt it" and "I've been given structured content like a compact/flattened JSON JWE, and will be expected to verify/decrypt it" but only one constructor that has to satisfy both.